File: lesson2/src/filesystem/AbstractFileSystemNode.java
package filesystem;

import java.util.Objects;

public abstract class AbstractFileSystemNode implements FileSystemNode {

    protected FileSystemNode parent;
    protected String name;

    protected AbstractFileSystemNode(String name) {
        this.name = Objects.requireNonNull(name, "Name is null");
    }

    @Override
    public final void setParent(FileSystemNode parent) {
        this.parent = parent;
    }

    @Override
    public final FileSystemNode getParent() {
        return parent;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public final String getPath() {
        return getParent() == null ? "/" + getName() : getParent().getPath() + "/" + getName();
    }
}

File: lesson3/src/filesystem/AbstractFileSystemNode.kt
package filesystem

abstract class AbstractFileSystemNode(
    override var nodeName: String = ""
) : FileSystemNode {

    final override var parent: FileSystemNode? = null

    final override fun path(): String {
        return if (parent == null) "/$nodeName" else parent!!.path() + "/$nodeName"
    }

}